dlt/destinations/impl/databricks/sql_client.py

```diff
@@ -0,0 +1,166 @@
+from contextlib import contextmanager, suppress
+from typing import Any, AnyStr, ClassVar, Iterator, Optional, Sequence, List, Union, Dict
+
+from databricks import sql as databricks_lib
+from databricks.sql.client import (
+    Connection as DatabricksSqlConnection,
+    Cursor as DatabricksSqlCursor,
+)
+from databricks.sql.exc import Error as DatabricksSqlError
+
+from dlt.common import pendulum
+from dlt.common import logger
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.destinations.exceptions import (
+    DatabaseTerminalException,
+    DatabaseTransientException,
+    DatabaseUndefinedRelation,
+)
+from dlt.destinations.sql_client import (
+    DBApiCursorImpl,
+    SqlClientBase,
+    raise_database_error,
+    raise_open_connection_error,
+)
+from dlt.destinations.typing import DBApi, DBApiCursor, DBTransaction
+from dlt.destinations.impl.databricks.configuration import DatabricksCredentials
+from dlt.destinations.impl.databricks import capabilities
+from dlt.common.time import to_py_date, to_py_datetime
+
+
+class DatabricksSqlClient(SqlClientBase[DatabricksSqlConnection], DBTransaction):
+    dbapi: ClassVar[DBApi] = databricks_lib
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, dataset_name: str, credentials: DatabricksCredentials) -> None:
+        super().__init__(credentials.catalog, dataset_name)
+        self._conn: DatabricksSqlConnection = None
+        self.credentials = credentials
+
+    def open_connection(self) -> DatabricksSqlConnection:
+        conn_params = self.credentials.to_connector_params()
+        self._conn = databricks_lib.connect(**conn_params, schema=self.dataset_name)
+        return self._conn
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+    @contextmanager
+    def begin_transaction(self) -> Iterator[DBTransaction]:
```

Comment:
we have a (seemingly currently unused) supports_transactions flag in the capabilities, why not use that in the superclass to determin wether transations are started here?

Reply:
and the following methods.

Reply:
Hmm yeah `supports_transactions` is only used in tests currently. Every db driver needs slightly different transaction implementation, so if we implement this in base class I think it makes sense to just do this:

```python
    def begin_transaction(self) -> ContextManager[DBTransaction]:
        if not self.capabilities.supports_transactions:
            yield self
        else:
            raise NotImplementedError(f"{type(self).__name__} does not implement begin_transaction")

    def commit_transaction(self) -> None:
        if not self.capabilities.supports_transactions:
            return
        raise NotImplementedError(f"{type(self).__name__} does not implement commit_transaction")

    def rollback_transaction(self) -> None:
        if not self.capabilities.supports_transactions:
            return
        raise NotImplementedError(f"{type(self).__name__} does not implement rollback_transaction")
```

Reply:
yes, that would be cool, maybe in a follow-up PR?